shared-libs/transitions/src/transitions/index.js

```diff
@@ -0,0 +1,331 @@
+   * equal to transition rev.  If revs are the same then transition just ran.
+   */
+  return Boolean(
+    change &&
+      !change.deleted &&
+      doc &&
+      !isRevSame(doc, info) &&
+      transition.filter(doc, info)
+  );
+};
+
+/*
+ * All transitions reference the same change.doc and work in series to
+ * apply changes to it.  A transition is free to make async calls but the next
+ * transition will only run after the previous transitions's callback is
+ * called.  This is a performance optimization that allows us to apply N
+ * transitions (updates) to a document with the cost of a single database
+ * change/write.
+ */
+const applyTransition = ({ key, change, transition }, callback) => {
+  if (!canRun({ key, change, transition })) {
+    logger.debug(
+      `canRun test failed on transition ${transition.key} for doc ${change.id} seq ${change.seq}`
+    );
+    return callback();
+  }
+
+  logger.debug(
+    `calling transition.onMatch for doc ${change.id} and transition ${key} seq ${change.seq}`
+  );
+
+  /*
+   * Transitions are async and should return true if the document has
+   * changed.  If a transition errors then log the error, but don't return it
+   * because that will stop the series and the other transitions won't run.
+   */
+  transition
+    .onMatch(change)
+    .then(changed => {
+      logger.debug(
+        `finished transition ${key} doc ${change.id} is ${changed ? 'changed' : 'unchanged'} seq ${change.seq}`
+      );
+      if (!changed) {
+        return changed;
+      }
+      return infodoc.updateTransition(change, key, true).then(() => {
+        return changed;
+      });
+    })
+    .catch(err => {
+      // adds an error to the doc but it will only get saved if there are
+      // other changes too.
+      const message = err.message || JSON.stringify(err);
+      utils.addError(change.doc, {
+        code: `${key}_error'`,
+        message: `Transition error on ${key}: ${message}`,
+      });
+      logger.error(`transition ${key} errored on doc ${change.id} seq ${change.seq}: ${JSON.stringify(err)}`);
+      if (!err.changed) {
+        return false;
+      }
+      return infodoc.updateTransition(change, key, false).then(() => {
+        return true;
+      });
+    })
+    .then(changed => callback(null, changed)); // return the promise instead
+};
+
+const processChange = (change, callback) => {
+  lineage
+    .fetchHydratedDoc(change.id)
+    .then(doc => {
+      change.doc = doc;
+      return infodoc.get(change).then(infoDoc => {
+        change.info = infoDoc;
+        // Remove transitions from doc since those
+        // will be handled by the info doc(sentinel db) after this
+        if (change.doc.transitions) {
+          delete change.doc.transitions;
+        }
+        applyTransitions(change, callback);
+      });
+    })
+    .catch(err => {
+      logger.error('transitions: fetch failed for %s : %o', change.id, err);
+      return callback(err);
+    });
+};
+
```

Comment:
This is new, and used in API to run transitions over a collection of docs, returning an updated list of docs that API will save as before.